File: AlgorithmicToolbox/week4_divide_and_conquer/1_binary_search/1_binary_search.cpp
#include <iostream>
#include <cassert>
#include <vector>
/*
In this problem, the goal is to implement the binary search algorithm that allows
searching very efficiently (even huge) lists, provided that the list is sorted.
*/
using std::vector;

int linear_search(const vector<int>& a, int x) {
    for (size_t i = 0; i < a.size(); ++i) {
        if (a[i] == x) return i;
    }
    return -1;
}

int binary_search(const vector<int>& a, int left, int right, int x) {
    if (right > left) {
        int mid = left + (right - left) / 2;
        if (a[mid] == x) {
            return mid;
        }
        else if (a[mid] > x) {
            binary_search(a, left, mid - 1, x);
        }
        else {
            binary_search(a, mid + 1, right, x);
        }
    }
    else {
        return -1;
    }
}


int main() {
    int n;
    std::cin >> n;
    vector<int> a(n);
    for (size_t i = 0; i < a.size(); i++) {
        std::cin >> a[i];
    }
    int m;
    std::cin >> m;
    vector<int> b(m);
    for (int i = 0; i < m; ++i) {
        std::cin >> b[i];
    }
    for (int i = 0; i < m; ++i) {
        std::cout << binary_search(a, 0, a.size()-1, b[i]) << ' ';
    }
}

File: AlgorithmicToolbox/week4_divide_and_conquer/1_binary_search/1_binary_search.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


# python3
"""
In this problem, the goal is to implement the binary search algorithm that allows searching
very efficiently (even huge) lists, provided that the list is sorted.
"""

def linear_search(keys, query):
    for i in range(len(keys)):
        if keys[i] == query:
            return i

    return -1


def binary_search(keys, low, high, query):

    if high >= low:
 
        mid = (high + low) // 2
 
        if keys[mid] == query:
            return mid
        elif keys[mid] > query:
            return binary_search(keys, low, mid - 1, query)
        else:
            return binary_search(keys, mid + 1, high, query)
 
    else:
        return -1

if __name__ == '__main__':
    n = int(input())
    input_keys = list(map(int, input().split()))
    k = int(input())
    input_queries = list(map(int, input().split()))

    for q in input_queries:
        print(binary_search(input_keys, 0, len(input_keys)-1, q), end=' ')


# In[ ]:




